service-registry/registry-etcd/src/main/java/org/apache/servicecomb/registry/etcd/EtcdDiscovery.java

```diff
@@ -0,0 +1,207 @@
+import org.apache.servicecomb.config.BootStrapProperties;
+import org.apache.servicecomb.foundation.common.utils.JsonUtils;
+import org.apache.servicecomb.registry.api.Discovery;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.env.Environment;
+
+import com.google.common.collect.Lists;
+
+import io.etcd.jetcd.ByteSequence;
+import io.etcd.jetcd.Client;
+import io.etcd.jetcd.KeyValue;
+import io.etcd.jetcd.Watch;
+import io.etcd.jetcd.kv.GetResponse;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.WatchOption;
+
+public class EtcdDiscovery implements Discovery<EtcdDiscoveryInstance> {
+
+  private Environment environment;
+
+  private String basePath;
+
+  private EtcdRegistryProperties etcdRegistryProperties;
+
+  private Client client;
+
+  private InstanceChangedListener<EtcdDiscoveryInstance> instanceChangedListener;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(EtcdDiscovery.class);
+
+  @Autowired
+  @SuppressWarnings("unused")
+  public void setEnvironment(Environment environment) {
+    this.environment = environment;
+  }
+
+  @Autowired
+  @SuppressWarnings("unused")
+  public void setEtcdRegistryProperties(EtcdRegistryProperties etcdRegistryProperties) {
+    this.etcdRegistryProperties = etcdRegistryProperties;
+  }
+
+  @Override
+  public String name() {
+    return EtcdConst.ETCD_REGISTRY_NAME;
+  }
+
+  @Override
+  public boolean enabled(String application, String serviceName) {
+    return environment.getProperty(String.format(EtcdConst.ETCD_DISCOVERY_ENABLED, application, serviceName),
+        boolean.class, true);
+  }
+
+  @Override
+  public List<EtcdDiscoveryInstance> findServiceInstances(String application, String serviceName) {
+
+    String prefixPath = basePath + "/" + application + "/" + serviceName;
+    SingletonManager.getInstance().computeIfAbsent(prefixPath, serName -> {
+      Watch watchClient = client.getWatchClient();
+      try {
+        ByteSequence prefixByteSeq = ByteSequence.from(prefixPath, Charset.defaultCharset());
+        watchClient.watch(
+            prefixByteSeq,
+            WatchOption.builder().withPrefix(prefixByteSeq).build(),
+            resp -> {
+              Map<String, EtcdDiscoveryInstance> instanceMap = SingletonManager.getInstance()
+                  .get("instance:" + prefixPath);
+              resp.getEvents().forEach(event -> {
+                String key = event.getKeyValue().getKey().toString(Charset.defaultCharset());
+                switch (event.getEventType()) {
+                  case PUT -> {
+                    EtcdDiscoveryInstance newInstance = getEtcdDiscoveryInstance(event.getKeyValue());
+                    instanceMap.put(key, newInstance);
+                  }
+                  case DELETE -> instanceMap.remove(key);
+                  default -> LOGGER.error("unkonw event");
```

Comment:
Is there an update?

Reply:
put and update are equal. They are all put in the event type.